### PR TITLE
remove Compare trait, add tests, make all lattice types PartialOrd, Eq, PartialEq

### PR DESCRIPTION
**Pull request by zzlk**

### CODE REVIEW

File: lattices/src/fake.rs [hunk 1]
```diff
@@ -32,15 +33,44 @@ impl<T> ConvertFrom<Fake<T>> for Fake<T> {
...
 }
 
+impl<T, O> PartialOrd<Fake<O>> for Fake<T>
+where
+    T: PartialOrd<O>,
+{
+    fn partial_cmp(&self, other: &Fake<O>) -> Option<std::cmp::Ordering> {
+        self.0.partial_cmp(&other.0)
+    }
+}
```
Comment (MingweiSamuel) [inline lattices/src/fake.rs:43]: This one seems dubious
I guess though any compare is consistent with the merge if the merge just panics

File: lattices/src/ord.rs [hunk 2]
```diff
@@ -41,18 +41,9 @@ impl<T> ConvertFrom<Max<T>> for Max<T> {
...
-{
-    fn compare(&self, other: &Max<T>) -> Option<std::cmp::Ordering> {
-        Some(Ord::cmp(&self.0, &other.0))
-    }
-}
-
 /// A totally ordered min lattice. Merging takes the smaller value.
 #[repr(transparent)]
-#[derive(Default, PartialEq, PartialOrd, Eq, Ord, Debug, Clone)]
+#[derive(Copy, Clone, Debug, Default, PartialOrd, Ord, PartialEq, Eq)]
```
Comment (MingweiSamuel) [inline lattices/src/ord.rs:46]: !!! This `PartialOrd, Ord` will be backwards. In the min lattice, 0 > 1

Could also maybe try `type Min<T> = Max<std:::cmp::Reverse<T>>` or something, not sure how ergonomic that is

File: lattices/src/pair.rs [hunk 3]
```diff
@@ -69,13 +71,52 @@ where
...
+    fn eq(&self, other: &Pair<LatAOther, LatBOther>) -> bool {
+        if self.a != other.a {
+            return false;
+        }
+
+        if self.b != other.b {
+            return false;
+        }
+
+        true
```
Comment (MingweiSamuel) [inline lattices/src/pair.rs:89]: Can just do `self.a == other.a && self.b == other.b` I think?

"a != b if and only if !(a == b)."

File: lattices/src/pair.rs [hunk 4]
```diff
@@ -69,13 +71,52 @@ where
...
+    }
+}
+impl<LatASelf, LatBSelf> Eq for Pair<LatASelf, LatBSelf>
+where
+    LatASelf: PartialEq,
+    LatBSelf: PartialEq,
```
Comment (MingweiSamuel) [inline lattices/src/pair.rs:95]: Should be `Eq` bounds. And can also just do `#[derive(Eq)]`

File: lattices/src/bottom.rs [hunk 5]
```diff
@@ -49,51 +46,52 @@ impl<Inner> ConvertFrom<Bottom<Inner>> for Bottom<Inner> {
...
+    fn eq(&self, other: &Bottom<Other>) -> bool {
+        match (&self.0, &other.0) {
+            (None, None) => true,
+            (None, Some(_)) => false,
+            (Some(_), None) => false,
+            (Some(this_inner), Some(other_inner)) => this_inner == other_inner,
+        }
+    }
+}
+impl<Inner> Eq for Bottom<Inner> where Inner: PartialEq {}
```
[the remark this answers is not in the document]
Comment (MingweiSamuel) [reply lattices/src/bottom.rs:76]: Also this should be `Inner: Eq`

File: lattices/src/dom_pair.rs [hunk 6]
```diff
@@ -68,27 +69,69 @@ where
...
+
+        true
+    }
+}
+impl<KeySelf, ValSelf> Eq for DomPair<KeySelf, ValSelf>
+where
+    KeySelf: PartialEq,
+    ValSelf: PartialEq,
+{
+}
```
[the remark this answers is not in the document]
Comment (MingweiSamuel) [reply lattices/src/dom_pair.rs:109]: And should be `Eq` bound not `PartialEq`

File: lattices/src/fake.rs [hunk 7]
```diff
@@ -32,15 +33,44 @@ impl<T> ConvertFrom<Fake<T>> for Fake<T> {
...
+
+impl<T, O> PartialEq<Fake<O>> for Fake<T>
+where
+    T: PartialEq<O>,
+{
+    fn eq(&self, other: &Fake<O>) -> bool {
+        self.0 == other.0
+    }
+}
+impl<T> Eq for Fake<T> where T: PartialEq {}
```
Comment (MingweiSamuel) [inline lattices/src/fake.rs:53]: Should be `T: Eq` or just `#[derive(Eq)]`

File: lattices/src/map_union.rs [hunk 8]
```diff
@@ -131,15 +133,60 @@ where
...
+
+        true
+    }
+}
+impl<MapSelf, K, ValSelf> Eq for MapUnion<MapSelf>
+where
+    MapSelf: Map<K, ValSelf, Key = K, Item = ValSelf> + MapIter + SimpleKeyedRef,
+    ValSelf: PartialEq,
+{
+}
```
Comment (MingweiSamuel) [inline lattices/src/map_union.rs:188]: `ValSelf: Eq`

could do `Self: PartialEq<Self>` for the complicated bounds I think

File: lattices/src/set_union.rs [hunk 9]
```diff
@@ -82,6 +83,21 @@ where
...
+{
+    fn eq(&self, other: &SetUnion<SetOther>) -> bool {
+        if self.0.len() != other.0.len() {
+            return false;
+        }
+
+        self.0.iter().all(|key| other.0.contains(&*key))
+    }
+}
+impl<SetSelf, Item> Eq for SetUnion<SetSelf> where SetSelf: Set<Item, Item = Item> + Iter {}
```
Comment (MingweiSamuel) [inline lattices/src/set_union.rs:99]: could do `Self: PartialEq<Self>` instead of the complicated bounds I think

**Review by MingweiSamuel — CHANGES_REQUESTED** [review]

File: lattices/src/test/mod.rs [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+    use std::cmp::Ordering::*;
+
+    for a in test_vec {
+        for b in test_vec {
+            for c in test_vec {
+                // Partial Eq:
+                // a != b if and only if !(a == b).
+                assert_eq!(a != b, !(a == b));
+
+                // Eq:
```
[the remark this answers is not in the document]
Comment (zzlk) [reply lattices/src/test/mod.rs:17]: I structured it like this because those comments/code chunks are copied from the PartialEq/PartialOrd documentation/wikipedia page.

[the remark this answers is not in the document]
Comment (MingweiSamuel) [reply lattices/src/test/mod.rs:17]: I dont feel strongly about it

File: lattices/src/ord.rs (hunk 2 above)
Comment (zzlk) [reply lattices/src/ord.rs:46]: nice catch :)

File: lattices/src/fake.rs (hunk 1 above)
Comment (zzlk) [reply lattices/src/fake.rs:43]: I think compare does need to be implemented for Fake to be useful at all, but it could also just panic immediately like merge does. I think I'll go with that implementation.

File: lattices/src/ord.rs (hunk 2 above)
Comment (zzlk) [reply lattices/src/ord.rs:46]: yeah, wrapping the reverse in there makes it pretty unergonomic because now you can't do Min::new(0);

File: lattices/src/pair.rs (hunk 3 above)
Comment (zzlk) [reply lattices/src/pair.rs:89]: yep, works.

File: lattices/src/map_union.rs (hunk 8 above)
Comment (zzlk) [reply lattices/src/map_union.rs:188]: yep, I think it works.

File: lattices/src/test/mod.rs (hunk 10 above)
[the remark this answers is not in the document]
Comment (zzlk) [reply lattices/src/test/mod.rs:17]: Will leave it as it is.

**Comment by MingweiSamuel** [comment]
Might also consider a marker trait to mark the lattice-specific semantics
```rust
pub trait LatticeOrd<Rhs>: PartialOrd<Rhs> {}
 ```

File: lattices/src/ord.rs (hunk 2 above)
Comment (MingweiSamuel) [reply lattices/src/ord.rs]: Its a shame

File: lattices/src/pair.rs [hunk 11]
```diff
@@ -52,30 +55,61 @@ where
...
-        let ord_b = self.b.compare(&other.b);
+    fn partial_cmp(&self, other: &Pair<LatAOther, LatBOther>) -> Option<Ordering> {
+        let ord_a = self.a.partial_cmp(&other.a);
+        let ord_b = self.b.partial_cmp(&other.b);
         if ord_a == ord_b {
             ord_a
         } else {
             None
         }
     }
```
Comment (MingweiSamuel) [inline lattices/src/pair.rs:72]: hmm I guess actually this is wrong Lol. Ill make a separate PR

**Review by MingweiSamuel — APPROVED** [review]

File: lattices/src/pair.rs (hunk 11 above)
Comment (MingweiSamuel) [reply lattices/src/pair.rs:72]: It was broken before this PR
